### PR TITLE
{2023.06}[foss/2023a] OpenFOAM v2312

### PR DESCRIPTION
**Pull request by boegel**
```
2 out of 113 required modules missing:

* KaHIP/3.16-gompi-2023a (KaHIP-3.16-gompi-2023a.eb)
* OpenFOAM/v2312-foss-2023a (OpenFOAM-v2312-foss-2023a.eb)
```

### CODE REVIEW

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.1-2023a.yml [hunk 1]
```diff
@@ -4,3 +4,4 @@ easyconfigs:
   - R-bundle-Bioconductor-3.18-foss-2023a-R-4.3.2.eb:
       options:
         from-pr: 20379
+  - OpenFOAM-v2312-foss-2023a.eb
```
Comment (ocaisa) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.1-2023a.yml:7]:
```suggestion
  - OpenFOAM-v2312-foss-2023a.eb
      options:
        include-easyblocks-from-pr: 3328
        from-pr: 20517
```

Comment (ocaisa) [reply easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.1-2023a.yml:7]: Dot com version was broken for users of `wmake`

**Comment by bedroge** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3

**Comment by bedroge** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3

**Comment by bedroge** [comment]
Sanity check is failing because of missing binaries (e.g. `boundaryFoam` and `engineFoam`). They should be in `OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/bin`, but they're not, while the same version on generoso does have them:
```
$ ls -1 /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/bin/ | wc -l
133

$ ls -1 /project/boegelbot/Rocky8/haswell/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/bin | wc -l
316
$ 
```

But I can't find any explanation for this in the logs...

**Comment by bedroge** [comment]
Oh, actually, it looks like the build step did fail, even though EB says it completed. In the log file of the build step itself I see lots of errors near the bottom, most of them are like this:
```
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: insertFeaturePoints.C:(.text+0xc19): undefined reference to `__gmpn_mul'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFO
AM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `mpfr_get_emin'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFO
AM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `__gmpn_com'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFO
AM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `__gmpq_swap'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFO
AM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `__gmpq_init'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `__gmpn_copyi'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `mpfr_set_q'
/cvmfs/software.eessi.io/versions/2023.06/compat/linux/x86_64/usr/bin/ld.bfd: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/amd/zen3/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libconformalVoronoiMesh.so: undefined reference to `__gmpn_add_n'

```

**Comment by ocaisa** [comment]
Ah, GMP is not included in the list of dependencies but it clearly is one

**Comment by bedroge** [comment]
> Ah, GMP is not included in the list of dependencies but it clearly is one

Yep, and MPFR.

**Comment by bedroge** [comment]
Doing an interactive build now with both of them added to the deps:
```
    ('GMP', '6.2.1'),
    ('MPFR', '4.2.0'),
```

It's worrying though that the build actually failed, but apparently this `Allwmake` command returned with an exit code 0?

**Comment by boegel** [comment]
> Doing an interactive build now with both of them added to the deps:
> 
> ```
>     ('GMP', '6.2.1'),
>     ('MPFR', '4.2.0'),
> ```
> 
> It's worrying though that the build actually failed, but apparently this `Allwmake` command returned with an exit code 0?

If I remember correctly this was one of the motivations for implementing the sanity check concept in EasyBuild :)

**Comment by bedroge** [comment]
Adding those dependencies didn't help, still getting the same issue/errors.

**Comment by bedroge** [comment]
It's probably related to CGAL, which can be built in different ways. We currently use a header-only version of CGAL, but when using this with OpenFOAM, it looks like MPFR and GMP are still required. There seems to be functionality in OpenFOAM that should detect which variant of CGAL you have, but I find lots of similar issues, e.g. this one with Spack: https://develop.openfoam.com/Development/openfoam/-/issues/3101.

**Comment by bedroge** [comment]
Confirmed, it's related to CGAL. In Spack they even stick to CGAL 4, probably because of similar issues:
https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/openfoam/package.py#L372

I suspect it pops up now because we're filtering $LD_LIBRARY_PATH. As a sort of a quick/dirty workaround I've added a command to `preconfigopts` that injects the MPFR and GMP directories to CGAL_LIBS somewhere, and that allowed the build to complete.

**Comment by bedroge** [comment]
https://github.com/easybuilders/easybuild-easyconfigs/pull/20841 and https://github.com/easybuilders/easybuild-easyblocks/pull/3366 should fix this.

**Comment by bedroge** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3

File: easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.1-2023a.yml [hunk 2]
```diff
@@ -22,3 +22,7 @@ easyconfigs:
         from-pr: 20784
   - Valgrind-3.21.0-gompi-2023a.eb
   - OrthoFinder-2.5.5-foss-2023a.eb
+  - OpenFOAM-v2312-foss-2023a.eb:
```
Comment (boegel) [inline easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.1-2023a.yml:25]: @bedroge Let's start using EasyBuild v4.9.2 for this?

Comment (bedroge) [reply easystacks/software.eessi.io/2023.06/eessi-2023.06-eb-4.9.1-2023a.yml:25]: Sure!

**Comment by bedroge** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by bedroge** [comment]
> New job on instance `eessi-bot-mc-aws` for architecture `x86_64-intel-skylake_avx512` for repository `eessi.io-2023.06-software` in job dir `/project/def-users/SHARED/jobs/2024.06/pr_563/13051`
> 
> date	job status	comment
> Jun 19 10:03:54 UTC 2024	submitted	job id `13051` awaits release by job manager
> Jun 19 10:04:28 UTC 2024	released	job awaits launch by Slurm scheduler
> Jun 19 10:12:56 UTC 2024	running	job `13051` is running
> Jun 19 10:45:32 UTC 2024	finished	
> 😢 FAILURE _(click triangle for details)_
> Jun 19 10:45:32 UTC 2024	test result	
> 😁 SUCCESS _(click triangle for details)_

```
make: /bin/sh: Argument list too long
make: *** [/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/intel/skylake_avx512/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/wmake/makefiles/general:210: /cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/intel/skylake_avx512/software/OpenFOAM/v2312-foss-2023a/OpenFOAM-v2312/platforms/linux64GccDPInt32Opt/lib/libOpenFOAM.so] Error 127
```

Looks like the path for skylake is too long...

**Comment by marconetto** [comment]
HI, just to double check, once this is issue is fixed, will we be able to run OpenFoam on zen4?

**Comment by ocaisa** [comment]
@marconetto Yes and no, the Zen4 builds are happening separately as we still have to bring that stack in line with the other architectures. Having said that, zen4 can't go through without getting this PR over the line. It's not an easy one though as this is a difficult issue and will require a little design

**Comment by marconetto** [comment]
thanks for the clarification @ocaisa

**Comment by bedroge** [comment]
Based on my comments in https://github.com/EESSI/software-layer/issues/296#issuecomment-2210439694, I was hoping to fix this by only unsetting some environment variables. That didn't work though, and I'm still not fully sure why. The failing command does work fine when running it interactively, also with the same environment variables set as during the build. My only guess is that there's a bunch of makefiles calling/including other ones, which are also somehow increasing the size of the environment and apparently exceeding some limit.

While analyzing and debugging this, I came up with a workaround though. The failing command is compiling `libOpenFOAM.so`, and this needs to link together a whole bunch of object files. It uses the full path to these object files, but since the command is running in the build/installation directory, it can use relative paths. This greatly reduces the length of the command. The following hook uses that approach and worked on skylake (still need to be cleaned/extended a bit):

```
def pre_build_hook_openfoam(self, *args, **kwargs):
    if self.name == 'OpenFOAM':
        objects_regex = (
            r'(\$E \$\(LINKLIBSO\) \$\(FOAM_EXTRA_LDFLAGS\)) \$\(OBJECTS\) (-L\$\(LIB_PLATFORMS\))',
            r'\1 $(subst $(WM_PROJECT_DIR),$(shell realpath --relative-to=$(PWD) $(WM_PROJECT_DIR)),$(OBJECTS)) \2'
        )
        apply_regex_substitutions('wmake/makefiles/general', [objects_regex])
```

**Comment by boegel** [comment]
@bedroge Awesome work!

Instead of implementing this in a hook, why don't we bake it into the OpenFOAM easyblock instead, since this should be a general fix, it's not specific to EESSI at all?

**Comment by bedroge** [comment]
> @bedroge Awesome work!
> 
> Instead of implementing this in a hook, why don't we bake it into the OpenFOAM easyblock instead, since this should be a general fix, it's not specific to EESSI at all?

Yes, you're absolutely right. I can open a PR in the easyblocks repo.

**Comment by bedroge** [comment]
I've opened an easyblocks PR: https://github.com/easybuilders/easybuild-easyblocks/pull/3388.
Let's give it a spin...

**Comment by bedroge** [comment]
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by ocaisa** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen4
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Comment by ocaisa** [comment]
bot: build repo:eessi.io-2023.06-software arch:aarch64/generic
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_n1
bot: build repo:eessi.io-2023.06-software arch:aarch64/neoverse_v1
bot: build repo:eessi.io-2023.06-software arch:x86_64/generic
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen2
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen3
bot: build repo:eessi.io-2023.06-software arch:x86_64/amd/zen4
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/haswell
bot: build repo:eessi.io-2023.06-software arch:x86_64/intel/skylake_avx512

**Review by trz42 — APPROVED** [review]
Looks good. The `zen4` build only contains one additional package. So, that stack is getting closer too.

**Review by ocaisa — APPROVED** [review]

**Pull request merged**